File: app/src/debug/java/com/bopr/android/smailer/ui/LongAsyncTask.java
package com.bopr.android.smailer.ui;

import android.app.Activity;
import android.os.AsyncTask;

import java.lang.ref.WeakReference;

/**
 * Task showing full screen infinite progress.
 *
 * @author Boris Pronin (<a href="mailto:boprsoft.dev@gmail.com">boprsoft.dev@gmail.com</a>)
 */
@SuppressWarnings("deprecation")
public abstract class LongAsyncTask<Params, Progress, Result> extends AsyncTask<Params, Progress, Result> {

    /* Holding activity reference in nested static class throws memory leak warning. This approach avoids it.*/
    private final WeakReference<Activity> activityReference;
    private android.app.ProgressDialog dialog;  /* do not replace qualifier with import to avoid deprecation warning */

    @SuppressWarnings("WeakerAccess")
    public LongAsyncTask(Activity activity) {
        this.activityReference = new WeakReference<>(activity);
    }

    @Override
    protected void onPreExecute() {
        dialog = new android.app.ProgressDialog(getActivity());
        dialog.setMessage("Processing...");
        dialog.show();
    }

    @Override
    protected void onPostExecute(Result result) {
        dialog.dismiss();
        super.onPostExecute(result);
    }

    public Activity getActivity() {
        return activityReference.get();
    }
}

File: app/src/debug/kotlin/com/bopr/android/smailer/ui/LongAsyncTask.kt
package com.bopr.android.smailer.ui

import android.app.ProgressDialog
import android.os.AsyncTask
import androidx.fragment.app.FragmentActivity
import java.lang.ref.WeakReference

/**
 * Task showing full screen infinite progress.
 *
 * @author Boris Pronin ([boprsoft.dev@gmail.com](mailto:boprsoft.dev@gmail.com))
 */
@Suppress("DEPRECATION")
abstract class LongAsyncTask<P, O, R>(activity: FragmentActivity) : AsyncTask<P, O, R>() {

    /* Holding activity reference in nested static class throws memory leak warning. This approach avoids it.*/
    private val activityReference: WeakReference<FragmentActivity> = WeakReference(activity)
    /* Do not replace qualifier with import to avoid deprecation warning */
    private lateinit var dialog: ProgressDialog

    val activity: FragmentActivity
        get() = activityReference.get()!!

    override fun onPreExecute() {
        dialog = ProgressDialog(activity)
        dialog.setMessage("Processing...")
        dialog.show()
    }

    override fun onPostExecute(result: R) {
        dialog.dismiss()
        super.onPostExecute(result)
    }

}